### PR TITLE
[BOLT] Add split function support for the Linux kernel

### PR DESCRIPTION
**Pull request by maksfb**
While rewriting the Linux kernel, we try to fit optimized functions into their original boundaries. When a function becomes larger, we skip it during the rewrite and end up with less than optimal code layout. To overcome that issue, add support for --split-function option so that hot part of the function could be fit into the original space. The cold part should go to reserved space in the binary.

### CODE REVIEW

File: bolt/lib/Passes/SplitFunctions.cpp [hunk 1]
```diff
@@ -715,6 +715,12 @@ Error SplitFunctions::runOnFunctions(BinaryContext &BC) {
   if (!opts::SplitFunctions)
     return Error::success();
 
+  if (BC.IsLinuxKernel && BC.BOLTReserved.empty()) {
+    BC.errs() << "BOLT-ERROR: split functions require reserved space in the "
```
Comment (dcci) [inline bolt/lib/Passes/SplitFunctions.cpp:719]: longer term idea/ drive-by comment: having to remember to type `BOLT-ERROR:` every time is cognitive load we might want to avoid. Potentially we might want at some point to do a global mark-and-sweep and introduce something like `llvm::bolt::bolt_error()`, or something along these lines.

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp [hunk 2]
```diff
@@ -248,6 +248,9 @@ class LinuxKernelRewriter final : public MetadataRewriter {
   /// Update ORC data in the binary.
   Error rewriteORCTables();
 
+  /// Validate written ORC tables after binary emission.
+  Error validateORCTables();
```
Comment (dcci) [inline bolt/lib/Rewrite/LinuxKernelRewriter.cpp:252]: If we can split the validation in its own commit, that would be good.

**Review by dcci — CHANGES_REQUESTED** [review]

File: bolt/lib/Passes/SplitFunctions.cpp (hunk 1 above)
Comment (maksfb) [reply bolt/lib/Passes/SplitFunctions.cpp:719]: Totally. Patches are welcome :)

File: bolt/lib/Rewrite/LinuxKernelRewriter.cpp (hunk 2 above)
Comment (maksfb) [reply bolt/lib/Rewrite/LinuxKernelRewriter.cpp:252]: Sure, if it helps you to review.

**Comment by maksfb** [comment]
FYI: moved new code for `BinaryContext.BOLTReserved` to #90766.

**Comment by dcci** [comment]
Good to go, both this one and #90766

**Review by dcci — APPROVED** [review]
